static/app/components/performance/searchBar.tsx

```diff
@@ -25,85 +26,172 @@ export type SearchBarProps = {
+      if (currentHighlightedItem) {
+        newSearchResults = getSearchGroupWithItemMarkedActive(
+          searchResults,
+          currentHighlightedItem,
+          false
+        );
       }
-      if (query.length < 3) {
-        setSearchResults([]);
+
+      if (nextHighlightedItem) {
+        newSearchResults = getSearchGroupWithItemMarkedActive(
+          newSearchResults,
+          nextHighlightedItem,
+          true
+        );
+      }
+
+      setSearchResults(newSearchResults);
+      return;
+    }
+
+    if (key === 'Enter') {
+      event.preventDefault();
+      const currentItem = searchResults[0].children[highlightedItemIndex];
+      if (!currentItem?.value) {
         return;
       }
-      setSearchString(query);
-      const projectIdStrings = (eventView.project as Readonly<number>[])?.map(String);
-      try {
-        setLoading(true);
-        const conditions = new MutableSearch('');
-        conditions.addFilterValues('transaction', [prepareQuery(query)], false);
-        conditions.addFilterValues('event.type', ['transaction']);
-
-        // clear any active requests
-        if (Object.keys(api.activeRequests).length) {
-          api.clear();
-        }
 
-        const useEvents = organization.features.includes(
-          'performance-frontend-use-events-endpoint'
-        );
-        const url = useEvents
-          ? `/organizations/${organization.slug}/events/`
-          : `/organizations/${organization.slug}/eventsv2/`;
-
-        const [results] = await doDiscoverQuery<{
-          data: Array<{'count()': number; project_id: number; transaction: string}>;
-        }>(api, url, {
-          field: ['transaction', 'project_id', 'count()'],
-          project: projectIdStrings,
-          sort: '-count()',
-          query: conditions.formatString(),
-          statsPeriod: eventView.statsPeriod,
-          referrer: 'api.performance.transaction-name-search-bar',
-        });
-
-        const parsedResults = results.data.reduce(
-          (searchGroup: SearchGroup, item) => {
-            searchGroup.children.push({
-              value: `${item.transaction}:${item.project_id}`,
-              title: item.transaction,
-              type: ItemType.LINK,
-              desc: '',
-            });
-            return searchGroup;
-          },
-          {
-            title: 'All Transactions',
-            children: [],
-            icon: null,
-            type: 'header',
+      handleSearch(currentItem.value);
+      return;
+    }
+  };
+
+  // `debounce` in functional components has to be wrapped in `useCallback`
+  // because otherwise the debounce is recreated on every render and becomes
+  // pointless. The exhaustive deps check is mad because it's not getting an
+  // inline function so it can't determine the dependencies. `useMemo` is a
+  // tempting replacement, but apparently React might randomly clear that to
+  // save memory
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
That makes sense! I usually put this stuff with the code since that's what people tend to read, but I can put it in the PR instead if that's better

Reply:
I usually prefer it since with a clean commit history (squashmerge) you have this way of explaining past decisions that doesn't litter the code base. It's a bit clean-code dogmatic but I usually do tend to stick to 'why' for comments with the occasional 'what' if something really funky is going on.

Reply:
😅 I agree with that, I just thought that this is both a "why" comment _and_ a "what" because something really funky is going on. To me, that the debounce here must be wrapped in a `useEffect` _and_ the dependency warning must be ignored is pretty funky! The commit history thing might break down in because as soon as someone fixes a typo in the comment, or re-indents the code the PR history is now near impossible to track down